src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java

```diff
@@ -15,247 +15,265 @@
+            .asText();
         if (user != null && !user.trim().isEmpty()) {
-            executionContext.setAttribute(ATTR_USER, user);
-            request.metrics().setUser(user);
+            ctx.setAttribute(ATTR_USER, user);
+            ctx.request().metrics().setUser(user);
         }
 
         // Extract scopes from introspection response
-        List<String> scopes = extractScopes(oauthResponseNode, oauth2.getScopeSeparator());
-        executionContext.setAttribute(ATTR_USER_ROLES, scopes);
+        List<String> scopes = extractScopes(oauthResponseNode, oauth2Resource.getScopeSeparator());
+        ctx.setAttribute(ATTR_USER_ROLES, scopes);
 
         // Check required scopes to access the resource
         if (oAuth2PolicyConfiguration.isCheckRequiredScopes()) {
             if (!hasRequiredScopes(scopes, oAuth2PolicyConfiguration.getRequiredScopes(), oAuth2PolicyConfiguration.isModeStrict())) {
-                sendError(
-                    OAUTH2_INSUFFICIENT_SCOPE_KEY,
-                    response,
-                    policyChain,
-                    "insufficient_scope",
-                    "The request requires higher privileges than provided by the access token."
-                );
-                return;
+                return sendError(ctx, OAUTH2_INSUFFICIENT_SCOPE_KEY, "insufficient_scope", INSUFFICIENT_SCOPES_MESSAGE);
             }
         }
 
         // Store OAuth2 payload into execution context if required
         if (oAuth2PolicyConfiguration.isExtractPayload()) {
-            executionContext.setAttribute(CONTEXT_ATTRIBUTE_OAUTH_PAYLOAD, oauth2payload);
+            ctx.setAttribute(CONTEXT_ATTRIBUTE_OAUTH_PAYLOAD, oauth2payload);
         }
 
         if (cacheResource != null) {
-            String accessToken = (String) executionContext.getAttribute(CONTEXT_ATTRIBUTE_OAUTH_ACCESS_TOKEN);
+            String accessToken = ctx.getAttribute(CONTEXT_ATTRIBUTE_OAUTH_ACCESS_TOKEN);
             CacheElement element = new CacheElement(accessToken, oauth2payload);
             if (oauthResponseNode.has(OAUTH_PAYLOAD_EXP)) {
                 long expTimestamp = oauthResponseNode.get(OAUTH_PAYLOAD_EXP).asLong();
                 long ttl = expTimestamp - System.currentTimeMillis() / 1000L;
                 element.setTimeToLive(Long.valueOf(ttl).intValue());
             }
-            cacheResource.getCache(executionContext).put(element);
+            cacheResource.getCache(ctx).put(element);
         }
 
         // Continue chaining
-        policyChain.doNext(request, response);
+        return Completable.complete();
+    }
+
+    private Completable introspectAccessToken(
+        RequestExecutionContext ctx,
+        String accessToken,
+        CacheResource<?> cacheResource,
+        OAuth2Resource<?> oauth2Resource
+    ) {
+        return Single
+            .<OAuth2Response>create(emitter -> oauth2Resource.introspect(accessToken, emitter::onSuccess))
```

Comment:
You didn't want to migrate Oauth2 resource to make it reactive ? ;)

Reply:
I was hopping you were brave enough to do it 😁
Definitely something we will have to do as things progress on jupiter. I'll add it to the list of the tasks 👍